tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/PoPTests.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Security;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Test.Common;
+using Microsoft.Identity.Test.LabInfrastructure;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Identity.Test.Integration.HeadlessTests
+{
+    // Currently PoP is supported only on .Net Classic
+#if DESKTOP
+    // Note: these tests require permission to a KeyVault Microsoft account;
+    // Please ignore them if you are not a Microsoft FTE, they will run as part of the CI build
+    [TestClass]
+    public class PoPTests
+    {
+        // This endpoint is hosted in the MSID Lab and is able to verify any pop token bound to an HTTP request
+        private const string PoPValidatorEndpoint = "https://signedhttprequest.azurewebsites.net/api/validateSHR";
+
+        private static readonly string[] s_scopes = { "User.Read" };
+
+        Uri protectedResournceUri1 = new Uri("https://www.contoso.com/path1/path2?queryParam1=a&queryParam2=b");
+        Uri protectedResournceUri2 = new Uri("https://www.bing.com/path3/path4?queryParam5=c&queryParam6=d");
+
+        // Doesn't exist, but the POP validator endpoint will check if the POP token matches this HTTP request 
+
+        private string _popValidationEndpointSecret;
+
+        [ClassInitialize]
+        public static void ClassInitialize(TestContext context)
+        {
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+        }
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            TestCommon.ResetInternalStaticCaches();
+            if (_popValidationEndpointSecret == null)
+            {
+                _popValidationEndpointSecret = LabUserHelper.KeyVaultSecretsProvider.GetSecret(
+                    "https://buildautomation.vault.azure.net/secrets/automation-pop-validation-endpoint/841fc7c2ccdd48d7a9ef727e4ae84325").Value;
+            }
+        }
+
+        [TestMethod]
+        public async Task POPAsync()
+        {
+            var labResponse = await LabUserHelper.GetDefaultUserAsync().ConfigureAwait(false);
+            await RunHappyPathTestAsync(labResponse).ConfigureAwait(false);
+        }
+
+        private async Task RunHappyPathTestAsync(LabResponse labResponse)
+        {
+            var user = labResponse.User;
```

Comment:
nit: is there a reason to make this a separate variable? instead of just labResponse.User.Upn , etc...?

Reply:
No, I guess not, it's just that typing `labResposne.User.property` everytime is kind of long...

Reply:
lol

---
In reply to: [352708502](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1525#discussion_r352708502) [](ancestors = 352708502)